### PR TITLE
poseidon struct with fieldhasher trait

### PR DESCRIPTION
**Pull request by GhostOfGauss**
Adds the Poseidon hash in arkworks-plonk-circuits.  Function operates on field elements only, not variables yet.  Adds new FieldHasher trait

### CODE REVIEW

**Comment by filiplazovic** [comment]
This is not a replacement for Poseidon PLONK Circuit. Instead, it's a different impl of native Poseidon https://github.com/webb-tools/arkworks-gadgets/blob/master/arkworks-gadgets/src/poseidon/mod.rs

**Comment by drewstone** [comment]
This LGTM

**Comment by filiplazovic** [comment]
I'll review this in the morning. I will have several comments

File: arkworks-gadgets/src/poseidon/field_hasher.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+trait FieldHasher<F: PrimeField> {
+	fn hash(&self, inputs: &Vec<F>) -> Result<F, PoseidonError>;
+	fn hash_two(&self, left: &F, right: &F) -> Result<F, PoseidonError>;
+}
+
+impl<F: PrimeField> FieldHasher<F> for Poseidon<F> {
+	fn hash(&self, inputs: &Vec<F>) -> Result<F, PoseidonError> {
+		let mut state = vec![F::zero()];
```
Comment (filiplazovic) [inline arkworks-gadgets/src/poseidon/field_hasher.rs:24]: The size of this state depends on the `width` from the parameters. (state.len() == self.params.width)

The first element is always zero. The rest of the parameters are inputs.
If the `inputs` length is less than `self.params.width - 1`, populate the rest of the state with zero elements.
If the `inputs` length is more than `self.params.width - 1`, return error.

File: arkworks-gadgets/src/poseidon/field_hasher.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+		for f in inputs {
+			state.push(*f);
+		}
+		let result = permute( &self.params, state)?;
```
Comment (filiplazovic) [inline arkworks-gadgets/src/poseidon/field_hasher.rs:28]: Please run `cargo fmt` locally, our CI check is not working

File: arkworks-gadgets/src/poseidon/field_hasher.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	
+	fn hash_two(&self, left: &F, right: &F) -> Result<F, PoseidonError> {
+		let state = vec![F::zero(), *left, *right];
+		let result = permute( &self.params, state)?;
+
+		Ok(result.get(0).cloned().ok_or(PoseidonError::InvalidInputs)?)
+	}
+}
+
+pub fn permute<F: PrimeField>(params: &PoseidonParameters<F>, mut state: Vec<F>)
```
Comment (filiplazovic) [inline arkworks-gadgets/src/poseidon/field_hasher.rs:41]: There is no reason for `permute` not to be part of `Poseidon` struct impl

**Comment by filiplazovic** [comment]
Can you add a test for checking if this implementation is the same as `CRH`?

**Review by filiplazovic — APPROVED** [review]

**Pull request merged**